### PR TITLE
tools/manifestfile: Add --unix-ffi option.

### PR DESCRIPTION
**Pull request by projectgus**
Allows adding unix-ffi library from the command line, to fix micropython-lib builds.

Follow up to 35dd959133fb233d75e9b3cddbf98b2ed01c6594 from https://github.com/micropython/micropython/pull/13620

This doesn't expose the full `add_library()` function to the command line, as currently the only use case is unix-ffi libraries. The only "external" user of this tool is currently micropython-lib.

Before:

```
 python ../../tools/manifestfile.py --lib . --compile unix-ffi/email.message/manifest.py
Error in manifest file: /home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.message/manifest.py: Package 're' not found in any known library.
```

After:

```
❯ python ../../tools/manifestfile.py --unix-ffi --lib . --compile unix-ffi/email.message/manifest.py
version=0.5.3 description=None license=None author=None pypi=None pypi_publish=None
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/ffilib/ffilib.py', target_path='ffilib.py', timestamp=1676863157.6312485, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d3a10>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/re/re.py', target_path='re.py', timestamp=1707800376.548602, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d3190>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/binascii/binascii.py', target_path='binascii.py', timestamp=1676863157.6112483, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dcc90>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/os/os/__init__.py', target_path='os/__init__.py', timestamp=1677138094.1277487, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86de150>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/os-path/os/path.py', target_path='os/path.py', timestamp=1707800376.548602, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dc9d0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/uu/uu.py', target_path='uu.py', timestamp=1676863157.6245818, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d3610>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/struct/struct.py', target_path='struct.py', timestamp=1676863157.617915, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dc910>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/base64/base64.py', target_path='base64.py', timestamp=1707800376.545269, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d3490>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/random/random.py', target_path='random.py', timestamp=1676863157.617915, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dc510>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/datetime/datetime.py', target_path='datetime.py', timestamp=1700610061.7551172, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dcbd0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/collections/collections/__init__.py', target_path='collections/__init__.py', timestamp=1677138094.1244152, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd7d0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/collections-defaultdict/collections/defaultdict.py', target_path='collections/defaultdict.py', timestamp=1677138094.1244152, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dda50>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/urllib.parse/urllib/parse.py', target_path='urllib/parse.py', timestamp=1680739294.753335, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dc090>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/warnings/warnings.py', target_path='warnings.py', timestamp=1676863157.6245818, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dc950>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/quopri/quopri.py', target_path='quopri.py', timestamp=1676863157.617915, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd290>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/functools/functools.py', target_path='functools.py', timestamp=1676863157.6112483, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd590>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/python-stdlib/string/string.py', target_path='string.py', timestamp=1676863157.617915, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86df0d0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.encoders/email/quoprimime.py', target_path='email/quoprimime.py', timestamp=1677138094.1344154, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd150>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.encoders/email/base64mime.py', target_path='email/base64mime.py', timestamp=1677138094.1344154, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd150>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.encoders/email/encoders.py', target_path='email/encoders.py', timestamp=1677138094.1344154, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd150>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.errors/email/errors.py', target_path='email/errors.py', timestamp=1677138094.1344154, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dc3d0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.charset/email/charset.py', target_path='email/charset.py', timestamp=1677138094.1344154, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86dd8d0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.utils/email/utils.py', target_path='email/utils.py', timestamp=1680739294.753335, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d28d0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.message/email/message.py', target_path='email/message.py', timestamp=1677138094.1377487, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d1bd0>, opt=None)
ManifestOutput(file_type=1, full_path='/home/gus/ry/george/micropython/lib/micropython-lib/unix-ffi/email.message/email/iterators.py', target_path='email/iterators.py', timestamp=1677138094.1377487, kind=6, metadata=<__main__.ManifestPackageMetadata object at 0x7a8cd86d1bd0>, opt=None)
```

*This work was funded through GitHub Sponsors.*

### CODE REVIEW

File: tools/manifestfile.py [hunk 1]
```diff
@@ -632,6 +633,8 @@ def main():
         exit(1)
 
     m = ManifestFile(mode, path_vars)
+    for lib in args.add_library:
+        m.add_library(lib, os.path.join("$(MPY_LIB_DIR)", lib))
```
Comment (dpgeorge) [inline tools/manifestfile.py:637]: I'm not sure it should prefix it with `MPY_LIB_DIR`.  What if the user wants a completely different path?

Comment (projectgus) [reply tools/manifestfile.py]: Have changed it so it takes a path instead.

File: tools/manifestfile.py [hunk 2]
```diff
@@ -603,6 +603,9 @@ def main():
         default=os.path.join(os.path.dirname(__file__), "../lib/micropython-lib"),
         help="path to micropython-lib repo",
     )
+    cmd_parser.add_argument(
```
Comment (jimmo) [inline tools/manifestfile.py:606]: Thanks @projectgus I think this is the right approach, but I think we might need the name and path to be both specified explicitly. Suggest we do do `name,path = path.split("=", 1)`

Also the use of `basename` below might be problematic to infer the name:

```
>>> import os
>>> os.path.basename("/foo/bar")
'bar'
>>> os.path.basename("/foo/bar/")
''
```

**Comment by dpgeorge** [comment]
I think it's fine to make this a special option called `--unix-ffi` that just does the right thing.

This tool running as a command-line tool is anyway an internal tool to MicroPython (only used by `micropython-lib` CI as far as I can tell), so we don't need to worry too much about changing/adding/removing these command-line options.

**Comment by projectgus** [comment]
@dpgeorge @jimmo I've updated this to add `--unix-ffi` only, with a short additional rationale in the description.

File: tools/manifestfile.py [hunk 3]
```diff
@@ -632,6 +635,8 @@ def main():
         exit(1)
 
     m = ManifestFile(mode, path_vars)
+    if args.unix_ffi:
+        m.add_library("unix-ffi", os.path.join("$(MPY_LIB_DIR)", "unix-ffi"))
```
Comment (dpgeorge) [inline tools/manifestfile.py:639]: This should use `m.add_library(..., prepend=True)`.

File: tools/manifestfile.py [hunk 4]
```diff
@@ -603,6 +603,9 @@ def main():
         default=os.path.join(os.path.dirname(__file__), "../lib/micropython-lib"),
         help="path to micropython-lib repo",
     )
+    cmd_parser.add_argument(
+        "--unix-ffi", action="store_true", help="Add dependency to unix-ffi libraries."
```
Comment (dpgeorge) [inline tools/manifestfile.py:607]: Maybe "prepend unix-ffi to the library path".

**Comment by jimmo** [comment]
Yep happy to make this just `--unix-ffi`. Thanks @projectgus

**Comment by jimmo** [comment]
(Although it might be useful for CI in a different non-micropython-lib repo, anyway we can deal with that in a future PR, let's just get the CI working for micropython-lib)

File: tools/manifestfile.py (hunk 3 above)
Comment (projectgus) [reply tools/manifestfile.py]: Oops, could have sworn I kept this. Good catch!

File: tools/manifestfile.py (hunk 4 above)
Comment (projectgus) [reply tools/manifestfile.py]: Done!

**Comment by projectgus** [comment]
> (Although it might be useful for CI in a different non-micropython-lib repo, which was the reason we added support for add_library initially. Anyway we can deal with that in a future PR, let's just get the CI working for micropython-lib)

I think it would be really good to have a good story on how to build and publish micropython packages outside of micropython-lib. Maybe I'm just missing the relevant knowledge, but it seems to me like actually building that out might result in a slightly different high-level design to what's currently here anyhow.

**Comment by dpgeorge** [comment]
Thanks for updating.  Now merged.

**Pull request merged**